net_tools: Introduce networkd

ssahani opened the pull request:
##### SUMMARY
This work is minimum set of configuration in aim of automating
systemd-networkd conf files.
Please see the doc section to how to configure yaml files.
More information  see https://www.freedesktop.org/software/systemd/man/systemd-networkd.service.html

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

Closes: #28709

##### COMPONENT NAME
networkd
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
"2.8 "
```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
Initial WIP Version
```
Signed-off-by: Susant Sahani <susant@redhat.com>

ssahani committed:
net_tools: Introduce networkd

This work is minimum set of configuration in aim of automatiting
systemd-networkd conf files.

Please see https://www.freedesktop.org/software/systemd/man/systemd-networkd.service.html

Signed-off-by: Susant Sahani <susant@redhat.com>

Closes: 28709

ryansb commented on lib/ansible/modules/net_tools/networkd.py:
```diff
@@ -0,0 +1,650 @@
+    - Configure Link, NetDev and Network for systemd-networkd
+
+options:
+  ConfigType:
```
The capitalization casing doesn't match what is typical in Ansible. We heavily prefer `snake_case` to `CamelCase` for module parameters, so mixing up snake/camel options isn't a source of frustration for users.

andreaso commented:
Wouldn't this new _networkd_ module rather belong in the `system/` namespace, which already contains the following system networky modules?

* firewalld
* hostname
* iptables
* ufw

ssahani commented:
It does the networking conf same ad nmcli . So I thought It should be better belongs there.

andreaso commented:
Good point! Guess the current namespace separation isn't without some inconsistencies.

ssahani committed:
Add more examples:

1. Add more examples
2. Rename join_bond to master bond etc.
3. improve NetworkdUtilities class

Signed-off-by: Susant Sahani <susant@redhat.com>

ssahani committed:
Cleanup: remove some variables as not used.

Signed-off-by: Susant Sahani <susant@redhat.com>

ssahani committed:
Add support for More tunnels:

Signed-off-by: Susant Sahani <susant@redhat.com>

ssahani committed:
Add support to configure vxlan

Add vxlan params.
Validate tunnel params

Signed-off-by: Susant Sahani <susant@redhat.com>

ssahani committed:
fix pep8 errors

ssahani committed:
Add support to configure multiple addresses.

ssahani committed:
Add support to configure routes

'''
       - networkd:
           config_type=network
           name=eth0
           routes="to=0.0.0.0/0 via=9.0.0.1 metric=100, to=192.168.5.0/24 via=192.168.5.1 metric=100"
           state=present

       - debug:
          var: result
          verbosity: 3

'''

Signed-off-by: Susant Sahani <susant@redhat.com>

ssahani commented:
Ping ! Any feedback Please.

maxamillion commented:
Hello and thank you for the pull request!

I'm curious, what's the goal of making this into a module vs creating a role and hosting it on [Ansible Galaxy](https://galaxy.ansible.com/)?

ssahani commented:
any reason why it should be there not in here. We have nmcli iproute stuff also here. I am trying to figure out why you think so ?

KellerFuchs commented:
@ssahani Hey, sorry your PR got kinda forgotten about.
If you are still up for getting that new module merged, I can make time for a review.  :)

KellerFuchs commented:
@maxamillion I think it makes sense to have it done as a module:
- There is fairly complex logic that would be complicated (and slow) to run as a playbook or role.
- You might want to call it several times in the execution of a playbook, with different arguments.
- You might want to notify handlers based on whether the module changed anything.

ssahani commented:
@KellerFuchs Sorry I don't have any time to work on this now. If you can rework and and merge that will be nice .

KellerFuchs commented:
@ssahani Unfortunately, I'm also a bit short on snoons and time currently.
I will see what I can do, but if you get around to working on this before I do, feel free to reopen the PR and ping me (@KellerFuchs in a Github comment should drop me a notification)
